# Previous

- CompletableFuture 는 비동기 작업의 예외를 처리할 수 있도록 `handle[Async]()` , `whenComplete[Async]()` , `exceptionally[Async]()` 를 제공한다.

# 💡 exceptionally[Async] (Function<Throwable, T> fn)

- 작업이 예외로 완료된 경우 해당 예외를 처리하고 새로운 결과나 대체 예외를 반환할 수 있다.
- 작업이 예외로 완료된 경우에만 실행되고 정상적으로 완료된 경우 이 함수는 실행되지 않는다.
- `Function<Throwable, T>` 형식의 함수를 인수로 받는다.
- 새로운 `CompletableFuture<T>` 객체를 반환하며 CompletableFuture 에 결과 또는 예외를 저장할 수 있다.

![image](https://github.com/shin-je-woo/TIL/assets/39439576/b32cc342-cc67-44bb-8c1f-2e00d6c6fbe2)

# 💡 handle[Async] (BiFunction<T, Throwable, U> fn)

- 결과와 예외 처리를 모두 다룰 수 있는 메서드로 비동기 작업의 예외 처리와 결과 처리를 동시에 수행할 때 유용하다.
- `BiFunction<T, Throwable, U>` 형식의 함수를 인수로 받으며 인수의 의미는 다음과 같다.
  - 첫 번째 인수는 비동기 작업이 성공적으로 완료된 경우 결과를 나타낸다.
  - 두 번째 인수는 비동기 작업이 예외로 완료된 경우 해당 예외를 나타낸다.
- 반환 값 새로운 `CompletableFuture<T>` 객체를 반환하며 CompletableFuture 에 정상 결과 또는 예외를 저장한다.

![image](https://github.com/shin-je-woo/TIL/assets/39439576/deb1363a-91d1-462c-aa1b-10eae9d49411)

# 💡 whenComplete[Async] (BiConsumer<T, Throwable> action)

- 결과와 예외 처리를 모두 다룰 수 있는 메서드로 비동기 작업의 예외 처리와 결과 처리를 동시에 수행할 때 유용하다.
- `BiConsumer<T, Throwable>` 형식의 함수를 인수로 받으며 인수의 의미는 다음과 같다.
  - 첫 번째 인수는 비동기 작업이 성공적으로 완료된 경우 결과를 나타낸다.
  - 두 번째 인수는 비동기 작업이 예외로 완료된 경우 해당 예외를 나타낸다.
- 반환 값 새로운 `CompletableFuture<T>` 객체를 반환하며 CompletableFuture 에 정상 결과 또는 예외를 저장한다.

![image](https://github.com/shin-je-woo/TIL/assets/39439576/8bb861c9-47ba-4bf5-aa00-60b3223e43f5)

# 💡 handle vs exceptionally vs whenComplete

![image](https://github.com/shin-je-woo/TIL/assets/39439576/d7a229db-08dc-4a39-bff6-a776e0c6ed36)
